spark_log_parser/extractor.py

```diff
@@ -0,0 +1,213 @@
+                    continue
+                if tf.isdir():
+                    continue
+
+                self._add_to_stats_and_verify(tf.size)
+
+                extract_path = extract_dir.joinpath(tf.name)
+                extract_path.parent.mkdir(parents=True, exist_ok=True)
+                with tar_file.extractfile(tf.name) as source, open(extract_path, "wb") as target:
+                    shutil.copyfileobj(source, target)
+
+                # Remove the archive if it was extracted to free up space
+                sub_paths = self._extract(extract_path)
+                if extract_path not in sub_paths:
+                    self._remove_from_stats(tf.size)
+                    extract_path.unlink()
+
+                paths += sub_paths
+
+        return paths
+
+    def _extract_zip(self, archive_path: Path, extract_dir: Path):
+        paths = []
+
+        with zipfile.ZipFile(archive_path) as zfile:
+            for zinfo in zfile.infolist():
+                if self._should_skip_file(zinfo.filename):
+                    continue
+                if zinfo.is_dir():
+                    continue
+
+                self._add_to_stats_and_verify(zinfo.file_size)
+
+                extract_path = extract_dir.joinpath(zinfo.filename)
+                extract_path.parent.mkdir(parents=True, exist_ok=True)
+                with zfile.open(zinfo.filename) as source, open(extract_path, "wb") as target:
+                    shutil.copyfileobj(source, target)
+
+                # Remove the archive if it was extracted to free up space
+                sub_paths = self._extract(extract_path)
+                if extract_path not in sub_paths:
+                    self._remove_from_stats(zinfo.file_size)
+                    extract_path.unlink()
+
+                paths += sub_paths
+
+        return paths
+
+    def _extract_gz(self, archive_path: Path, extract_path: Path):
+        with open(archive_path, "rb") as fobj:
+            fobj.seek(-4, 2)
+            size = struct.unpack("I", fobj.read(4))[0]
+
+        self._add_to_stats_and_verify(size)
+
+        extract_path.parent.mkdir(parents=True, exist_ok=True)
+        with gzip.open(archive_path, "rb") as source, open(extract_path, "wb") as target:
+            shutil.copyfileobj(source, target)
+
+        # Remove the archive if it was extracted to free up space
+        sub_paths = self._extract(extract_path)
+        if extract_path not in sub_paths:
+            self._remove_from_stats(size)
+            extract_path.unlink()
+
+        return sub_paths
+
+    FILE_SKIP_PATTERNS = [".DS_Store".lower(), "__MACOSX".lower(), "/."]
+
+    def _should_skip_file(self, filename: str):
+        if filename.startswith("."):
+            return True
+
+        filename = filename.lower()
+        for name in Extractor.FILE_SKIP_PATTERNS:
+            if name in filename:
+                return True
+
+        return False
+
+    def _add_to_stats_and_verify(self, size, count=1):
```

Comment:
I like these methods for tracking the starts, I'm going to pirate this for the spark_predictor!